### PR TITLE
(Download tab) Enable generic assay download by selected entities

### PR DESCRIPTION
**Pull request by dippindots**
Fix cBioPortal/cbioportal#7851

This pr can only let users to download generic assay data in single study query since heatmap feature is only enabled for single study. Will extended to multiple query in the future.

[Example download tab link(without heatmap selection)](https://deploy-preview-3506--cbioportalfrontend.netlify.app/results/download?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=lusc_tcga_pan_can_atlas_2018&case_set_id=lusc_tcga_pan_can_atlas_2018_cnaseq&data_priority=0&gene_list=TP53%2520EGFR%2520KRAS&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=lusc_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=lusc_tcga_pan_can_atlas_2018_mutations&profileFilter=0&tab_index=tab_visualize)
[Example download tab link(with heatmap selection)](https://deploy-preview-3506--cbioportalfrontend.netlify.app/results/download?Action=Submit&RPPA_SCORE_THRESHOLD=2.0&Z_SCORE_THRESHOLD=2.0&cancer_study_list=lusc_tcga_pan_can_atlas_2018&case_set_id=lusc_tcga_pan_can_atlas_2018_cnaseq&data_priority=0&gene_list=TP53%2520EGFR%2520KRAS&generic_assay_groups=lusc_tcga_pan_can_atlas_2018_microbiome_signature%2CPrasinovirus%2CSfi1unalikevirus%2CSimplexvirus&geneset_list=%20&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=lusc_tcga_pan_can_atlas_2018_gistic&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=lusc_tcga_pan_can_atlas_2018_mutations&heatmap_track_groups=lusc_tcga_pan_can_atlas_2018_microbiome_signature%2CPrasinovirus%2CSfi1unalikevirus%2CSimplexvirus&profileFilter=0&tab_index=tab_visualize)

Describe changes proposed in this pull request:
- Select generic assay entities from OncoPrint tab then download from download tab
when no selection:
image
after select:
image
download file example:
image

- New utils function for getting generic assay meta property clearly `getGenericAssayMetaPropertyOrDefault`
- Unit tests for `getGenericAssayMetaPropertyOrDefault` function

### CODE REVIEW

**Comment by dippindots** [comment]
@jjgao @alisman, please use the link in the description for testing. Could you give some suggestions about the text I use when there is no selection: I use `Only entities selected in OncoPrint tab can be download at here. Please select entities first.` currently.

**Comment by inodb** [comment]
@dippindots maybe the text can be something like:

`Heatmap tracks added in the [oncoprint tab](link to oncoprint tab) can be downloaded here`

**Review by jjgao — APPROVED** [review]
LGTM

File: src/pages/resultsView/ResultsViewPageStore.ts [hunk 1]
```diff
@@ -610,6 +613,10 @@ export class ResultsViewPageStore {
         return getMolecularProfiles(this.urlWrapper.query);
     }
 
+    public handleTabChange(id: string, replace?: boolean) {
+        this.urlWrapper.updateURL({}, `results/${id}`, false, replace);
+    }
```
Comment (dippindots) [inline src/pages/resultsView/ResultsViewPageStore.ts:618]: moved from `ResultsViewPage.tsx` to `ResultsViewPageStore.ts` to reduce redundancy.

File: src/pages/resultsView/ResultsViewPageStore.ts [hunk 2]
```diff
@@ -4024,14 +4049,36 @@ export class ResultsViewPageStore {
...
+                        const sampleMolecularIdentifiers = _.flatten(
+                            _.map(molecularIds, molecularId =>
+                                _.map(this.samples.result, sample => {
+                                    return {
+                                        molecularProfileId: molecularId,
+                                        sampleId: sample.sampleId,
+                                    } as SampleMolecularIdentifier;
+                                })
+                            )
+                        );
```
Comment (dippindots) [inline src/pages/resultsView/ResultsViewPageStore.ts:4061]: Previous implementation will always download the whole sample list which is not correct. We should always download data for current selected samples.

File: src/shared/lib/GenericAssayUtils/GenericAssayCommonUtils.ts [hunk 3]
```diff
@@ -198,14 +203,29 @@ export function makeGenericAssayOption(
...
+    defaultValue: string = NOT_APPLICABLE_VALUE
+): string {
+    if (property in meta.genericEntityMetaProperties) {
+        return (meta.genericEntityMetaProperties as {
+            [property: string]: string;
+        })[property];
+    } else {
+        return defaultValue;
+    }
+}
```
Comment (dippindots) [inline src/shared/lib/GenericAssayUtils/GenericAssayCommonUtils.ts:231]: Create a getter function to make getting generic assay property more readable.

**Comment by alisman** [comment]
@dippindots not sure i understand: prior to selection, how do you get the text for the leftmost column?  "Microbiome signatures .."

**Comment by alisman** [comment]
@jjgao @dippindots @inodb does it make sense to ask user to go to oncoprint to select heatmap track?  that feels kinda cludgy to me.    could we not just give them dropdown for gene, dropdown for profile and then let them make selection right here?

**Comment by dippindots** [comment]
@alisman my answer to the first question: we can get the left column text because we can find the description of the `Generic Assay` profiles.

File: src/pages/resultsView/ResultsViewPageStore.ts [hunk 4]
```diff
@@ -4024,14 +4049,36 @@ export class ResultsViewPageStore {
...
 
-                        return fetchGenericAssayDataByStableIdsAndMolecularIds(
-                            stableIds,
-                            molecularIds
-                        ).then(genericAssayData => {
-                            genericAssayDataGroupByProfileIdSuffix[
-                                profileIdSuffix
-                            ] = genericAssayData;
-                        });
+                        const sampleMolecularIdentifiers = _.flatten(
```
Comment (alisman) [inline src/pages/resultsView/ResultsViewPageStore.ts:4052]: i think you can use _.flatMap to get rid of one level of nesting here

File: src/pages/resultsView/ResultsViewPageStore.ts (hunk 1 above)
Comment (alisman) [reply src/pages/resultsView/ResultsViewPageStore.ts:618]: where is the redundancy?

Comment (dippindots) [reply src/pages/resultsView/ResultsViewPageStore.ts:618]: Download tab cannot use a private function from `ResultsViewPage.tsx`, so this function need to be copied into `ResultsViewPageStore.ts`, so I just moved it to `ResultsViewPageStore.ts`.

**Comment by dippindots** [comment]
@alisman updated

**Review by alisman — APPROVED** [review]

**Pull request merged**